File: ruby/bob/bob.rb
class Bob
  def self.hey(remark)
    remark = remark.strip.chomp

    if yelling?(remark)
      return "Calm down, I know what I'm doing!" if remark[-1] == "?"
      "Whoa, chill out!"
    elsif remark[-1] == "?"
      "Sure."
    elsif remark == ""
      "Fine. Be that way!"
    else
      "Whatever."
    end
  end

  private
    def self.yelling?(remark)
      remark == remark.upcase && remark != remark.downcase
    end
end


module BookKeeping
  VERSION = 2
end

File: python/bob/bob.py
import string

def hey(phrase):
    phrase = ' '.join(phrase.split())
    upcase = phrase.upper()
    if phrase == "":
        return "Fine. Be that way!"
    elif phrase[-1] == "?":
        if phrase == upcase and has_letters(upcase):
            return "Calm down, I know what I'm doing!"
        else:
            return "Sure."
    elif phrase == upcase and has_letters(upcase):
        return "Whoa, chill out!"
    else:
        return "Whatever."

def has_letters(phrase):
    return len(set(phrase) - set(string.ascii_uppercase)) < len(set(phrase))
